spec/Candid.md

```diff
@@ -882,201 +882,150 @@ service { <name> : <functype>; <methtype>;* } <: service { <name> : <functype'>;
-<v> ~> <v'> : <t'>
---------------------------------------
-opt (<v> : <t>) ~> opt <v'> : opt <t'>
+c[null <: opt <t>](null) = null
 ```
 
-If an optional value does not have a subtype of the expected type, the result is `null`:
+An optional value coerces at an option type, if the constituent value has a suitable type, and else goes to `null`:
 ```
-not (<t> <: <t'>)
-----------------------------------
-opt (<v> : <t>) ~> null : opt <t'>
+c[opt <t> <: opt <t'>](null) = null
+c[opt <t> <: opt <t'>](opt <v>) = opt c[<t> <: <t'>](v)  if <t> <: <t'>
+c[opt <t> <: opt <t'>](opt <v>) = null                   if not(<t> <: <t'>)
 ```
 
-NOTE: The two rules above imply that a Candid decoder has to be able to decide the subtyping relation at runtime.
+NOTE: These rules above imply that a Candid decoder has to be able to decide the subtyping relation at runtime.
 
 Coercing a non-null, non-optional and non-reserved value at an option type treats it as an optional value:
 ```
-<v> ≠ null
-<v> ≠ (null : reserved)
-<v> ≠ opt _
-not (null <: <t>)
-opt <v> ~> <v'> : opt <t>
--------------------------
-<v> ~> <v'> : opt <t>
+c[<t> <: opt <t'>](<v>) = opt c[<t> <: <t'>](v)  if not (null <: <t>) and <t> <: <t'>
+c[<t> <: opt <t'>](_)   = null                   if not (null <: <t>) and not (<t> <: <t'>)
+c[reserved <: opt <t'>](_) = null
 ```
 
 #### Records
 
-Only records coerce at record type. Missing fields of option or reserved type turn into `null`.
+In the following rule:
+
+ * The `<nat1>*` field names are those present in both the actual and expected type. The values are coerced.
+ * the `<nat2>*` field names those only in the actual type. The values are ignored.
+ * the `<nat3>*` field names are those only in the expected type, which therefore must be of optional or reserved type. The `null` value is used for these.
 
-In the following rule, the `<nat1>*` field names are those present in both the value and the type, the `<nat2>*` field names only those in the value, and `<nat3>*` are those only in the type.
 ```
-<v1> ~> <v1'> : <t1>
-opt empty <: <t3>
----------------------------------------------------------------------------------------------------------------------------------------
-record { <nat1> = <v1>;* <nat2> = <v2>;* } ~> record { <nat1> = <v1'>;* <nat3> = null;* } : record {  <nat1> = <t1>;* <nat3> = <t3>;* }
+c[record { <nat1> = <t1>;* <nat2> = <t2>;* } <: record { <nat1> = <t1'>;* <nat3> = <t3>;* }](record { <nat1> = <v1>;* <nat2> = <v2>;* })
+    = record { <nat1> = c[<t1> <: <t1'>](<v1>);* <nat3> = null;* }
 ```
 
-
 #### Variants
 
 Only a variant value with an expected tag coerces at variant type.
 
 ```
-<v> ~> <v'> : <t>
-----------------------------------------------------------------------------------
-variant { <nat> = <v> } ~> variant { <nat> = <v'> } : variant { <nat> = <t>; _;* }
+c[variant { <nat> = <t>;* _;* } <: variant { <nat> = <t'>;* _;* }](variant { <natN> = <v> })
+    = variant { <natN> = c[<tN> <: <t'N>](<v>) }
```

Comment:
This formulation is a bit unclear, and perhaps more complicated than necessary. I assume the first `*` are the common tags, whereas the `_;*` are additional ones? But then, why can the latter occur on the l.h.s.? And how does `natN` relate to `nat*`?
```suggestion
c[variant { <nat> = <t>; _;* } <: variant { <nat> = <t'>; _;* }](variant { <nat> = <v> })
    = variant { <nat> = c[<t> <: <t'>](<v>) }
```

Reply:
Oh, yes, that's a bit bogus. Should be
```suggestion
c[variant { <nat> = <t>;* } <: variant { <nat> = <t'>;* _;* }](variant { <natN> = <v> })
    = variant { <natN> = c[<tN> <: <t'N>](<v>) }
```
where only the supertype has extra tags.

The `<natN>` is one of the `<nat>*`, I wasn’t sure how to best express that in this pseudo-syntax. `<nat>[i]` maybe?

Reply:
What's wrong with my simpler suggestion?

Reply:
Hmm, I guess I tried to keep a closer connection to the the subtyping rule, modeling that the supertype has more tags than the subtype… but you are right, it’s not needed here.